### PR TITLE
Move login flows

### PR DESCRIPTION
**Pull request by derek-globus**
[sc-35752](https://app.shortcut.com/globus/story/35752)

1. Move LoginFlowManager & its concrete implementations (CommandLine/LocalServer) out of experimental into `globus_sdk.login_flows`.
1. Add a reference docs page for public elements, updating some docstrings along the way.
1. Rename LocalServer-type errors & adjust the error hierarchy. 

----
📚 Documentation preview 📚: https://globus-sdk-python--1057.org.readthedocs.build/en/1057/

### CODE REVIEW

**Review by kurtmckee — CHANGES_REQUESTED** [review]

File: src/globus_sdk/login_flows/local_server_login_flow_manager/local_server.py [hunk 1]
```diff
@@ -33,19 +36,50 @@
...
+
 
 class RedirectHTTPServer(HTTPServer):
     """
-    HTTPServer that accepts an html_template to be displayed to the user
+    An HTTPServer which accepts a html_template to be displayed to the user
     and uses a Queue to receive an auth_code from its RequestHandler.
     """
 
+    WAIT_TIMEOUT = timedelta(minutes=5)
```
[the remark this answers is not in the document]
Comment (derek-globus) [reply src/globus_sdk/login_flows/local_server_login_flow_manager/local_server.py:81]: Oh I misinterpreted the timeout as 5 minutes. I thought I was clarifying not changing 🤦 

I'll update it to be 15 minutes.

File: changelog.d/20240918_120618_derek_move_login_flows.rst [hunk 2]
```diff
@@ -0,0 +1,8 @@
+
+Changed
+~~~~~~~
+
+- LoginFlowManagers have been moved from ``globus_sdk.experimental.login_flow_managers``
+  to ``globus_sdk.login_flows``. (:pr:`NUMBER`)
```
[the remark this answers is not in the document]
Comment (derek-globus) [reply changelog.d/20240918_120618_derek_move_login_flows.rst:6]: Discussed offline.

Basic synopsis. The problem between "flows" vs "login flows" is just a part of the problem of overloaded terms existing at globus. Hopefully the text here is explanatory enough to identify to users what we mean but if explanatory text isn't enough, we need to move the whole company away from using the term "login flow" (or change the name of one of our services).

Either way there isn't much to be done here. If we switch to "login flow managers" then "where's login timer managers?"

File: src/globus_sdk/login_flows/local_server_login_flow_manager/local_server.py [hunk 3]
```diff
@@ -33,19 +36,50 @@
...
+                ).encode("utf-8")
+            )
+
+            self.server.return_code(LocalServerLoginError(msg))
+
 
 class RedirectHTTPServer(HTTPServer):
     """
-    HTTPServer that accepts an html_template to be displayed to the user
+    An HTTPServer which accepts a html_template to be displayed to the user
```
[the remark this answers is not in the document]
Comment (derek-globus) [reply src/globus_sdk/login_flows/local_server_login_flow_manager/local_server.py:77]: Looked it up & I'm wrong
https://editorsmanual.com/articles/indefinite-article-a-an-with-abbreviations/

Switching back to "an"

File: src/globus_sdk/login_flows/command_line_login_flow_manager.py [hunk 4]
```diff
@@ -128,6 +122,11 @@ def print_authorize_url(self, authorize_url: str) -> None:
     def prompt_for_code(self) -> str:
         """
         Prompt the user to enter an authorization code.
+
+        This method is publicly exposed to allow for simpler customization through
+        subclassing and overriding.
+
+        :returns str: The authorization code entered by the user.
```
[the remark this answers is not in the document]
Comment (derek-globus) [reply src/globus_sdk/login_flows/command_line_login_flow_manager.py:129]: I will test and remove if so.

File: docs/authorization/login_flows.rst [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+environments (e.g., while ssh-ed into a cluster node).
+
+.. warning::
+
+    Globus Auth requires that redirect URIs, including the local server used
+    here, be pre-registered with the client in use.
+
+    Before using this flow, navigate to the
+    `Globus Developers Pane <https://auth.globus.org/v2/web/developers>`_ and ensure
+    that `https://localhost` is listed as an allowed "Redirect URL" for your client.
```
[the remark this answers is not in the document]
Comment (derek-globus) [reply docs/authorization/login_flows.rst:75]: I copied over my existing client's configured endpoint allowance.

Will circle back to this after I clean up the cosmetic changes you requested.

File: src/globus_sdk/login_flows/command_line_login_flow_manager.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (derek-globus) [reply src/globus_sdk/login_flows/command_line_login_flow_manager.py:129]: I'll remove

File: docs/authorization/login_flows.rst (hunk 5 above)
[the remark this answers is not in the document]
Comment (derek-globus) [reply docs/authorization/login_flows.rst:75]: I've started down this hole and cannot get back to a working state....

[the remark this answers is not in the document]
Comment (derek-globus) [reply docs/authorization/login_flows.rst:75]: Okay, this turned into a slack thread in #auth ([ref](https://globus.slack.com/archives/C03N0A8J7/p1726850494134019)).

Apparently the localserver behvaior (loopback url like localhost with an arbitrary port) is only supported for native clients and they don't need to allowlist it; it's supported by default.

With that knowledge, I'll remove this warning & add a client-side check that the client passed through to "LocalServerLoginFlowManager.for_globus_app(...)" must be a native client.

File: src/globus_sdk/login_flows/local_server_login_flow_manager/local_server_login_flow_manager.py [hunk 6]
```diff
@@ -132,20 +116,25 @@ def for_globus_app(
...
-        :param config: A ``GlobusAppConfig`` to configure the login flow.
-        :returns: A ``LocalServerLoginFlowManager`` instance.
-        :raises: GlobusSDKUsageError if a custom login_redirect_uri is defined in
-            the config.
+        Create a ``LocalServerLoginFlowManager`` for use in a GlobusApp.
+
+        :param app_name: The name of the app. Will be prefilled in native auth flows.
+        :param login_client: A client used to make Globus Auth API calls.
+        :param config: A GlobusApp-bounded object used to configure login flow manager.
+        :raises: GlobusSDKUsageError if app config is incompatible with the manager.
```
Comment (kurtmckee) [inline src/globus_sdk/login_flows/local_server_login_flow_manager/local_server_login_flow_manager.py:124]: Tagging this as an example to change but there may be others (or, we can do a mass-update after this merges).

```suggestion
        :raises GlobusSDKUsageError: if app config is incompatible with the manager.
```

**Review by kurtmckee — CHANGES_REQUESTED** [review]

**Review by kurtmckee — APPROVED** [review]

**Pull request merged**